x-pack/plugins/security_solution/public/detection_engine/rule_monitoring/components/basic/filters/multiselect_filter/index.tsx

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+import { noop } from 'lodash';
+import { EuiPopover, EuiFilterGroup, EuiFilterButton, EuiFilterSelectItem } from '@elastic/eui';
+
+export interface MultiselectFilterProps<T = unknown> {
+  dataTestSubj?: string;
+  title: string;
+  items: T[];
+  itemsSelectedByDefault?: T[];
+  onSelectionChange?: (selectedItems: T[]) => void;
+  renderItem?: (item: T) => React.ReactChild;
+  renderLabel?: (item: T) => string;
+}
+
+export const MultiselectFilterFactory = <T extends unknown>() => {
```

Comment:
Discussed with @xcrzx. I'll try to make the component itself generic, but it looks like there could be issues with types and `displayName`. We'll see if it works out or not.

Reply:
@xcrzx I managed to do it like that, but it still required type casting, unfortunately:

```ts
// We have to wrap it in a function and cast to original type because React.memo
// returns a component type which is not generic.
const enhanceMultiselectFilterComponent = () => {
  const Component = React.memo(MultiselectFilterComponent);
  Component.displayName = 'MultiselectFilter';
  return Component as typeof MultiselectFilterComponent;
};

export const MultiselectFilter = enhanceMultiselectFilterComponent();
```

I'm ok with either of the two approaches.